src/mbgl/renderer/render_source.cpp

```diff
@@ -64,12 +64,39 @@ void RenderSource::setObserver(RenderSourceObserver* observer_) {
 
 void RenderSource::onTileChanged(Tile& tile) {
     observer->onTileChanged(*this, tile.id);
+    if (tile.isComplete()) {
+        const auto tileID = tile.id.toUnwrapped();
+        const auto& idealTile = idealTilesLoaded.find(tileID);
+        if (idealTile != idealTilesLoaded.end()) {
+            idealTilesLoaded[tileID] = true;
+        }
+        for (const auto& entry : idealTilesLoaded) {
+            if (entry.second == false) {
+                return;
+            }
+        }
+        observer->onIdealTilesLoaded(*this);
+    }
 }
 
 void RenderSource::onTileError(Tile& tile, std::exception_ptr error) {
     observer->onTileError(*this, tile.id, error);
 }
 
+void RenderSource::onIdealTilesChanged(const std::vector<UnwrappedTileID>& idealTiles) {
```

Comment:
maybe it would be simpler, if we just introduced `const std::set<UnwrappedTileID>& TilePyramid::getIdealTiles() const `

Reply:
Done